various constructing bean improvement

radist-nt opened the pull request:
Please consider including this PR to your PR for lightbend/config. It includes:
* public check during constructor discovery
* more accurate exception handling
* support for single constructor compiled with `-parameters`

Sorry, didn't managed to setup sbt, so I've added new test in java and translated it to scala by hand. It could contain errors.

radist-nt committed:
various constructing bean improvement

* public check
* more accurate exception handling
* support for single constructor compiled with -parameters

radist-nt committed:
fixed hasPropertiesBindedToConstructor

lodik committed:
fix test compilation, add one test case

lodik commented:
Thank you for your contribution. This is useful functional. I was thinking about it when doing https://github.com/lightbend/config/pull/643, but decided to stop on @ConstructorProperties support.
Yes, your test contains few compilation errors, but I have fixed them in 4773a3e30a94c30f12c9811ae91edba0e9894441 in this repo. Also, I have added one case, which don't pass tests. Probably it's because of https://github.com/lodik/config/blob/master/config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java#L236

If you are a Docker user, you can easily run sbt in docker. Just execute 
`docker run --rm -it -v $(pwd):/app -w /app mozilla/sbt sbt "testOnly com.typesafe.config.impl.ConfigBeanFactoryTest"` in project's directory. It will compile the project and execute all tests in `ConfigBeanFactoryTest`. To execute all just use `sbt test`. Or if you are an IntellijIdea user, you can install the Scala plugin, import project from existing sources and IntellijIdea will care about all.

radist-nt committed:
fix test compiled without -parameters argument

radist-nt commented:
>  Also, I have added one case, which don't pass tests. Probably it's because of https://github.com/lodik/config/blob/master/config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java#L236

Seems like it's due to the compilation without parameter names. I've changed the test to work in both modes (with or without "-parameters" compiler argument), but to test both scenario tests should be run twice: the first time without -parameters compiler argument, the second time with it. To compile with parameters, add "-parameters" to https://github.com/lodik/config/blob/master/build.sbt#L70. Don't know whether it possible to force sbt to compile library second time just for testing.

I've installed IntelliJ IDEA and Scala plugin, but it didn't managed to extract the project structure from sbt. But I've tested the similar java code.

lodik commented:
I've fixed one compilation issue in the tests again (scala has not `:` operator, so you must use `if ... else ...`) and merged this pr manually after squashing. 
I want to update docs and then try to ping maintainers about review.
